server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -122,30 +113,66 @@ public TemporalAccessor parse(String input) {
         }
 
         try {
-            return parser.parse(input);
+            return doParse(input);
         } catch (DateTimeParseException e) {
             throw new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + format + "]", e);
         }
     }
 
+    private TemporalAccessor doParse(String input) {
+        if (parsers.size() > 1) {
+            for (DateTimeFormatter formatter : parsers) {
+                if (tryParseUnresolved(formatter, input) == true) {
+                    return formatter.parse(input);
+                }
+            }
+        }
+        return firstParser().parse(input);
+    }
+
+    /**
+     * Attempt parsing the input without throwing exception. This is needed because java-time requires ordering on optional (composite)
+     * patterns. Joda does not suffer from this.
+     * https://bugs.openjdk.java.net/browse/JDK-8188771
+     *
+     * @param input An arbitrary string resembling the string representation of a date or time
+     * @return true if parsing was successful, false if parsing failed
+     */
+    private boolean tryParseUnresolved(DateTimeFormatter formatter, String input) {
+        try {
+            ParsePosition pp = new ParsePosition(0);
+            formatter.parseUnresolved(input, pp);
+            int len = input.length();
+            if (pp.getErrorIndex() == -1 && pp.getIndex() == len) {
+                return true;
+            }
+        } catch (RuntimeException ex) {
+            // should not happen, but ignore if it does
+        }
+        return false;
+    }
+
     @Override
     public DateFormatter withZone(ZoneId zoneId) {
         // shortcurt to not create new objects unnecessarily
-        if (zoneId.equals(parser.getZone())) {
+        if (zoneId.equals(firstParser().getZone())) {
             return this;
         }
 
-        return new JavaDateFormatter(format, printer.withZone(zoneId), roundupParser.withZone(zoneId), parser.withZone(zoneId));
+        return new JavaDateFormatter(format, printer.withZone(zoneId),
+            parsers.stream().map(p->p.withZone(zoneId)).toArray(size->new DateTimeFormatter[size]));
     }
 
     @Override
     public DateFormatter withLocale(Locale locale) {
+        //TODO not sure that we can keep that shortcut, and the one above
```

Comment:
probably not because previously we had a common locale for the composite parser and now it is possible that each one has a different locale (not sure whether we should allow this though and throw an `IllegalArgumentException` in the constructor if that it is the case?) Also, we should probably still keep this optimization in case we only have one parser I think

Reply:
agree. I don't think we have a code were we define composite formatters with different `Locale`
We could also refactor this into verifying all parsers having the same `Locale` or `Zone` before returning this.